### DISCUSSION TITLE
Why am I getting LNK2038: mismatch detected for 'RuntimeLibrary'?

### DISCUSSION THREAD

**Discussion by asfarley** [microsoft/vcpkg#41344] [General]
I have a Windows MFC C++ application solution with several projects contained in the solution, developed in VS 2022. Some of the projects yield static libraries which are used by the  top-level application.

I'd like to use the 'cpr' library via vcpkg but I'm having some difficulty. I've tried following some guides for both non-Visual Studio integrated, and using the integrated vcpkg. For now, I'd like to focus on using the integrated package just to keep things as simple as possible (as far I know, I have no need for anything unusual).

I've now removed VCPKG_ROOT from my path in an attempt to prevent confusion between the built-in vcpkg and the Visual Studio version. 

In Visual Studio, one of the top-level application projects is failing to compile with LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MTd_StaticDebug' in project.obj	project	C:\XXXXX\cpr.lib(cprtypes.cpp.obj)	1		

I can see at some level what this means: my project is compiled as multi-thread static debug, and the cpr library presumably is compiled as Multithreaded DLL in dynamic release mode, which are incompatible. What I don't understand is why is is happening. 

I do have a vcpkg.json file in the root folder of my solution, with the contents:
```
{ 
"dependencies": [
    "cpr"
  ]
}

```

I also have a file vcpkg-configuration.json containing the following:
```
{
  "default-registry": {
    "kind": "git",
    "baseline": "6f1ddd6b6878e7e66fcc35c65ba1d8feec2e01f8",
    "repository": "https://github.com/microsoft/vcpkg"
  },
  "registries": [
    {
      "kind": "artifact",
      "location": "https://github.com/microsoft/vcpkg-ce-catalog/archive/refs/heads/main.zip",
      "name": "microsoft"
    }
  ]
}

```

A few questions I couldn't answer from the docs: If I'm using vcpkg to compile a static library project.lib with external dependency cpr.lib (presumably provided by vcpkg), do the other packages requiring project.lib also have to use vcpkg or should they point to the compiled pcr.lib file without worrying about/using vcpkg?

Should I have one vcpkg.json in the solution root folder or one per project folder?

How does vcpkg (in Visual Studio) choose a Runtime Library to target when it compiles a .lib in vcpkg.json?

I tried enabling vcpkg + manifest in my top-level application where the build is failing, and I get this:
![image](https://github.com/user-attachments/assets/b337810d-3b3c-4df1-97bb-29930486c530)

Please let me know what additional info would be helpful.

**Comment by asfarley** [comment]
Here's the vcpkg settings I have for all the projects:

![image](https://github.com/user-attachments/assets/6b7de01e-59f1-4057-9a7c-3caaa43690e9)

**Comment by asfarley** [comment]
It seems I'm going backwards. Previously I was able to compile some of my static libraries, and linking only failed when the top-level application was trying to pull in the libraries. 

Now, even my static libraries won't compile because 'vcpkg install' is failing:
```
1>------ Build started: Project: libVSLS, Configuration: Debug Win32 ------
1>Installing vcpkg dependencies to C:\VitalSim\vsls\vcpkg_installed\x86-windows\
1>"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\vcpkg\vcpkg.exe" install  --x-wait-for-lock --triplet "x86-windows" --vcpkg-root "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\vcpkg\\" "--x-manifest-root=C:\VitalSim\vsls\\" "--x-install-root=C:\VitalSim\vsls\vcpkg_installed\x86-windows\\"
1>Fetching registry information from https://github.com/microsoft/vcpkg (HEAD)...
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\vcpkg\scripts\buildsystems\msbuild\vcpkg.targets(183,5): warning : error: failed to fetch ref HEAD from repository https://github.com/microsoft/vcpkg
1>failed to execute: "C:\Program Files\Git\cmd\git.exe" "--git-dir=C:\Users\alexa\AppData\Local\vcpkg\registries\git\.git" "--work-tree=C:\Users\alexa\AppData\Local\vcpkg\registries\git" -c core.autocrlf=false fetch --update-shallow -- https://github.com/microsoft/vcpkg HEAD
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\vcpkg\scripts\buildsystems\msbuild\vcpkg.targets(183,5): warning : error: git failed with exit code: (128).
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\vcpkg\scripts\buildsystems\msbuild\vcpkg.targets(183,5): warning : error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\vcpkg\scripts\buildsystems\msbuild\vcpkg.targets(183,5): warning : error: 40 bytes of body are still expected
1>fetch-pack: unexpected disconnect while reading sideband packet
1>fatal: early EOF
1>fatal: fetch-pack: invalid index-pack output
```

I can't understand what's going on with the 'failed to fetch ref HEAD' error.

**Comment by asfarley** [comment]
Hmm, it appears that more than one thing might be going on here and the issue above could be due to network connectivity.

**Comment by asfarley** [comment]
Ok, now I'm back to the original issue: runtime library mismatch. The previous error was due to poor network connectivity.

**Comment by asfarley** [comment]
A bit more information: 
 * All my projects (static libs and top-level applications) use ISO C++ 17
 * Visual Studio 2022 v17.11.4
 * Windows 11
 * MSVC 19.41.34120 for x86
 * All projects have Runtime Library: Multi-threaded Debug (/MTd)

**Comment by Osyotr** [comment]
The CRT in your project and in libraries provided by vcpkg (usually) should match.
If your project uses static CRT, you should use `*-windows-static` triplet. If your project uses dynamic CRT, you should use `*-windows` or `*-windows-static-md`.
Based on the info from your screenshot you're using `x86-windows` (aka dynamic libs, dynamic CRT).
Try switching to `x86-windows-static` (aka static libs, static CRT).
